README.md
# BIP66

<p align="center">
    <img src="https://github.com/sleepdefic1t/BIP66/raw/master/bip66_logo.png" width=50% height=auto/>
</p>

> A simple Bitcoin BIP66 Implementation in C++ for the ARK Ecosystem.

ANS.1 DER Encoder/Decoder based on [Bitcoin BIP66](https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki).

[![GitHub Actions](https://github.com/sleepdefic1t/BIP66/workflows/GitHub%20Actions/badge.svg)](https://github.com/sleepdefic1t/BIP66)
[![CodeFactor](https://www.codefactor.io/repository/github/sleepdefic1t/bip66/badge)](https://www.codefactor.io/repository/github/sleepdefic1t/bip66)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/3828f000e03d45ba86997addd00834c0)](https://www.codacy.com/manual/sleepdefic1t/BIP66)
[![Latest Version](https://badgen.now.sh/github/release/sleepdefic1t/BIP66)](https://github.com/sleepdefic1t/BIP66/releases)
[![License: MIT](https://badgen.now.sh/badge/license/MIT/green)](https://opensource.org/licenses/MIT)

- [Supported Platforms](#supported-platforms)
- [Installation](#installation)
  - [Arduino IDE](#arduino-ide)
  - [PlatformIO](#platformio)
  - [CMake](#cmake)
- [Usage](#usage)
  - [Encoding a Signature](#encoding-a-signature)
  - [Decoding a Signature](#decoding-a-signature)
  - [Checking an Encoded Signature](#checking-an-encoded-signature)
- [License](#license)

## Supported Platforms

| OS:       | Supported:    |
| :--       | :--------:    |
| Linux     | ✅             |
| macOS     | ✅             |
| Windows   | ✅             |

| Environment:  | Supported:    |
| :-------      | :--------:    |
| Arduino IDE   | ✅             |
| PlatformIO    | ✅             |

| Board:    | Supported:    |
| :------   | :--------:    |
| ESP8266   | ✅             |
| ESP32     | ✅             |

## Installation

### Arduino IDE

Download BIP66 from the Arduino IDE Library manager.  
You can also copy this repo to the Arduino Libraries directory.  
(e.g. `~/Documents/Arduino/libraries`)

1) Open the `BIP66.ino` sketch in the `examples/BIP66` folder.  
2) Select your board from the Arduino IDE
3) Upload the sketch.

---

### PlatformIO

#### Building BIP66 with PIO

`pio run`

#### Building BIP66 and Tests with PIO

`pio run -t test/`

#### Flashing and Running Tests with PIO

`pio run -t test/ -e esp32 -t upload`

---

### CMake

Operating System builds like Linux, macOS, and Windows use CMake.  
This build should be done out of source.

#### Building BIP66 with CMake

1) `mkdir build && cd build`
2) `cmake ..`
3) `cmake --build .`

#### Building BIP66 and Tests with CMake

1) `mkdir build && cd build`
2) `cmake -DUNIT_TEST=ON ..`
3) `cmake --build .`

#### Running BIP66 Tests with CMake

`./test/bip66_tests`

---

## Usage

### Encoding a Signature

From the raw R & S elements:

```cpp
uint8_t r[32] = {
    33, 112, 79, 42, 219, 46, 74, 16, 163, 221, 193, 215, 214, 69, 82, 184, 6, 28, 5, 246, 209, 42, 22, 140, 105, 9, 28, 117, 88, 29, 97, 20
};

uint8_t s[32] = {
    14, 223, 55, 104, 157, 39, 134, 252, 105, 10, 249, 240, 246, 250, 31, 98, 156, 149, 105, 80, 57, 246, 72, 166, 212, 85, 72, 67, 2, 64, 46, 147
};

uint8_t signature[72];

const bool success = bip66::encode(r, 32, s, 32, signature);
```

---

### Decoding a Signature

From a DER-encoded signature to raw R & S elements:

```cpp
uint8_t derEncodedSignature[70] = {
    48, 68, 2, 32, 33, 112, 79, 42, 219, 46, 74, 16, 163, 221, 193, 215, 214, 69, 82, 184, 6, 28, 5, 246, 209, 42, 22, 140, 105, 9, 28, 117, 88, 29, 97, 20, 2, 32, 14, 223, 55, 104, 157, 39, 134, 252, 105, 10, 249, 240, 246, 250, 31, 98, 156, 149, 105, 80, 57, 246, 72, 166, 212, 85, 72, 67, 2, 64, 46, 147
};

uint8_t r[32];
uint8_t s[32];

const bool success = bip66::decode(derEncodedSignature, 70, r, s);
```

---

### Checking an Encoded Signature

From a DER-encoded signature:

```cpp
const uint8_t derEncodedSignature[70] = {
    48, 68, 2, 32, 33, 112, 79, 42, 219, 46, 74, 16, 163, 221, 193, 215, 214, 69, 82, 184, 6, 28, 5, 246, 209, 42, 22, 140, 105, 9, 28, 117, 88, 29, 97, 20, 2, 32, 14, 223, 55, 104, 157, 39, 134, 252, 105, 10, 249, 240, 246, 250, 31, 98, 156, 149, 105, 80, 57, 246, 72, 166, 212, 85, 72, 67, 2, 64, 46, 147
};

const bool success = bip66::check(derEncodedSignature, 70);
```

---

## License

[MIT](LICENSE) © [Simon Downey](https://github.com/sleepdefic1t)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.3.2][] - 2019-11-27

### Changed

- improvements to reduce code reuse and improve accuracy ([0.3.2][])

## [v0.3.1][] - 2019-10-08

### Changed

- improved overall implementation ([v0.3.1][])

## [0.2.1][] - 2019-05-20

### Changed

- exclude test folder from package export

## [0.2.0][] - 2019-02-13

### Added

- ported hex-string test-cases from `bitcoinjs` lib to byte/uint8_t-vectors.

### Changed

- change test name.
- replace Hex String test-cases with byte/uint8_t-vectors.
- improve testing overall (more test cases, testing valid/invalid).
- update Arduino Sketch to reflect changes.
- drop use of `hex.h` helper file.

## [0.1.0][] - 2019-02-09

### Added

- initial release ([0.1.0][])

[0.1.0]: https://github.com/sleepdefic1t/bip66/releases/tag/0.1.0
[0.2.0]: https://github.com/sleepdefic1t/bip66/compare/0.1.0...0.2.0
[0.2.1]: https://github.com/sleepdefic1t/bip66/compare/0.2.0...0.2.1
[v0.3.1]: https://github.com/sleepdefic1t/bip66/compare/0.2.0...v0.3.1
[0.3.2]: https://github.com/sleepdefic1t/bip66/compare/0.2.0...0.3.2


test/bip66.cpp
/********************************************************************************
*
*  The MIT License (MIT)
*
*  Copyright (c) Simon Downey <simon@ark.io> <https://github.com/sleepdefic1t>
*
*  Permission is hereby granted, free of charge, to any person obtaining a copy of
*  this software and associated documentation files (the "Software"), to deal in
*  the Software without restriction, including without limitation the rights to
*  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
*  the Software, and to permit persons to whom the Software is furnished to do so,
*  subject to the following conditions:
*
*  The above copyright notice and this permission notice shall be included in all
*  copies or substantial portions of the Software.
*
*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
*  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
*  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
*  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
*  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*
*  BIP66: https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
*  adapted from: https://github.com/bitcoinjs/bip66/blob/master/index.js
*
*********************************************************************************/

#include "gtest/gtest.h"

#include <bip66.hpp>

#include <array>
#include <cstdint>

#include "test_cases.hpp"

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, encode) {
    const uint8_t count     = ENC_DEC_COUNT;
    uint8_t matches         = 0U;

    std::array<uint8_t, ENCODED_LENGTH> signature {};

    while (++matches < count &&
           bip66::encode(test_cases::decoded[matches],
                         ELEMENT_LENGTH,
                         &test_cases::decoded[matches][ELEMENT_LENGTH],
                         ELEMENT_LENGTH,
                         signature.data())) {};

  ASSERT_EQ(matches, count);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, encode_valid) {
    const uint8_t count     = VALID_ENCODED_COUNT;
    uint8_t matches         = 0U;

    std::array<uint8_t, ENCODED_MAX_LENGTH> signature {};

    while (++matches < count &&
           bip66::encode(&test_cases::valid::r[matches][1],
                         test_cases::valid::r[matches][0],
                         &test_cases::valid::s[matches][1],
                         test_cases::valid::s[matches][0],
                         signature.data())) {};

    ASSERT_EQ(matches, count);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, encode_invalid) {
    const uint8_t count     = INVALID_RS_COUNT;
    uint8_t misses          = 0U;

    std::array<uint8_t, ENCODED_MAX_LENGTH> signature {};

    while (++misses < count &&
           !bip66::encode(&test_cases::invalid::r[misses][1],
                          test_cases::invalid::r[misses][0],
                          &test_cases::invalid::s[misses][1],
                          test_cases::invalid::s[misses][0],
                          signature.data())) {};

    ASSERT_EQ(misses, count);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, decode) {
    uint8_t matches = 0U;

    std::array<uint8_t, DECODED_LENGTH> rsBuffer {};

    while (bip66::decode(test_cases::encoded[++matches],
                         ENCODED_LENGTH,
                         rsBuffer.data(),
                         &rsBuffer[ELEMENT_LENGTH])) {};

    ASSERT_EQ(matches, ENC_DEC_COUNT);
}
////////////////////////////////////////////////////////////////////////////////

TEST(BIP66, decode_valid) {
    const uint8_t count     = VALID_ENCODED_COUNT;
    uint8_t matches         = 0U;

    std::array<uint8_t, DECODED_LENGTH + 1U> rsBuffer {};

    while (++matches < count &&
           bip66::decode(&test_cases::valid::encoded[matches][1],
                         test_cases::valid::encoded[matches][0],
                         rsBuffer.data(),
                         &rsBuffer[ELEMENT_LENGTH])) {};

    ASSERT_EQ(matches, count);
}
////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, decode_invalid) {
    const uint8_t count     = VALID_ENCODED_COUNT;
    uint8_t misses          = 0U;

    std::array<uint8_t, DECODED_LENGTH> rsBuffer {};

    while (++misses < count &&
           !bip66::decode(&test_cases::invalid::encoded[misses][1],
                          test_cases::invalid::encoded[misses][0],
                          rsBuffer.data(),
                          &rsBuffer[ELEMENT_LENGTH])) {};

    ASSERT_EQ(misses, count);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, check) {
    uint8_t matches = 0U;

    while (bip66::check(test_cases::encoded[++matches], ENCODED_LENGTH)) {};

    ASSERT_EQ(matches, ENC_DEC_COUNT);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, check_valid) {
    const uint8_t count     = VALID_ENCODED_COUNT;
    uint8_t matches         = 0U;

    while (++matches < count &&
           bip66::check(&test_cases::valid::encoded[matches][1],
                        test_cases::valid::encoded[matches][0])) {};

    ASSERT_EQ(matches, VALID_ENCODED_COUNT);
}

////////////////////////////////////////////////////////////////////////////////
TEST(BIP66, check_invalid) {
    const uint8_t count     = INVALID_ENCODED_COUNT;
    uint8_t misses          = 0U;

    while (++misses < count &&
           !bip66::check(test_cases::invalid::encoded[misses],
                         test_cases::invalid::encoded[misses][0])) {};

    ASSERT_EQ(misses, count);
}


test/iot/main.cpp
/********************************************************************************
*
*  The MIT License (MIT)
*
*  Copyright (c) Simon Downey <simon@ark.io> <https://github.com/sleepdefic1t>
*
*  Permission is hereby granted, free of charge, to any person obtaining a copy of
*  this software and associated documentation files (the "Software"), to deal in
*  the Software without restriction, including without limitation the rights to
*  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
*  the Software, and to permit persons to whom the Software is furnished to do so,
*  subject to the following conditions:
*
*  The above copyright notice and this permission notice shall be included in all
*  copies or substantial portions of the Software.
*
*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
*  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
*  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
*  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
*  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*
*  BIP66: https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
*  adapted from: https://github.com/bitcoinjs/bip66/blob/master/index.js
*
*********************************************************************************/

#if (defined PLATFORMIO)

#if (defined UNIT_TEST)

#include <AUnit.h>

void setup() {
  Serial.begin(115200);

  aunit::TestRunner::setTimeout(0);

  delay(1000);
}

void loop() {
  aunit::TestRunner::run();
}

#else

#include <Arduino.h>

void setup() {}
void loop() {}

#endif

#endif


src/bip66.h
/**
 * bip66.h
 * 
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 *
 * author: Simon Downey <simon@ark.io>
 *  
 * BIP66: https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
 * adapted from: https://github.com/bitcoinjs/bip66/blob/master/index.js
 **/

#ifndef BIP66_H
#define BIP66_H

#include <cstddef>
#include <cstdint>
#include <vector>

struct BIP66 {
  static constexpr size_t ELEMENT_LEN = 32U;
  static constexpr size_t SIG_MIN_LEN = 8U;
  static constexpr size_t SIG_MAX_LEN = 72U;

  static bool encode(const std::vector<uint8_t>& r,
                     const std::vector<uint8_t>& s,
                     std::vector<uint8_t>& outSignature);

  static bool encode(const uint8_t* rawSignature,
                     std::vector<uint8_t>& outSignature);

  static bool decode(const std::vector<uint8_t>& signature,
                     std::vector<uint8_t>& outR,
                     std::vector<uint8_t>& outS);

  static bool check (const std::vector<uint8_t>& signature);
  static bool check(const uint8_t* signature, size_t len);
};

#endif


src/bip66.hpp
/********************************************************************************
*
*  The MIT License (MIT)
*
*  Copyright (c) Simon Downey <simon@ark.io> <https://github.com/sleepdefic1t>
*
*  Permission is hereby granted, free of charge, to any person obtaining a copy of
*  this software and associated documentation files (the "Software"), to deal in
*  the Software without restriction, including without limitation the rights to
*  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
*  the Software, and to permit persons to whom the Software is furnished to do so,
*  subject to the following conditions:
*
*  The above copyright notice and this permission notice shall be included in all
*  copies or substantial portions of the Software.
*
*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
*  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
*  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
*  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
*  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*
*  BIP66: https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
*  adapted from: https://github.com/bitcoinjs/bip66/blob/master/index.js
*
*********************************************************************************/

#ifndef BIP66_SRC_HPP
#define BIP66_SRC_HPP

#include <cstdint>

namespace bip66 {

////////////////////////////////////////////////////////////////////////////////
bool encode(const uint8_t *rElement,
            const uint8_t &rElementLength,
            const uint8_t *sElement,
            const uint8_t &sElementLength,
            uint8_t *outSignature);

////////////////////////////////////////////////////////////////////////////////
bool decode(const uint8_t *signature,
            const uint8_t &signatureLength,
            uint8_t *outR,
            uint8_t *outS);

////////////////////////////////////////////////////////////////////////////////
bool check(const uint8_t *signature, const uint8_t &signatureLength);

}  // namespace bip66

#endif


examples/BIP66/hex.h
/**
 * This file is part of Ark Cpp Crypto.
 *
 * (c) Ark Ecosystem <info@ark.io>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 * 
 * this file is from: https://github.com/ArkEcosystem/cpp-crypto/blob/master/src/include/cpp-crypto/helpers/encoding/hex.h
 **/

#ifndef HEX_H
#define HEX_H

#include <string>
#include <vector>

/**
 * Hex Helpers
 **/
template <typename T>
inline std::string BytesToHex(const T itbegin, const T itend, bool fSpaces = false) {
  std::string rv;
  static const char hexmap[16] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f'};
  rv.reserve((itend - itbegin) * 3);
  for (T it = itbegin; it < itend; ++it) {
    const auto val = static_cast<uint8_t>(*it);
    if (fSpaces && it != itbegin) {
      rv.push_back(' ');
    }
    rv.push_back(hexmap[val >> 4]);
    rv.push_back(hexmap[val & 15]);
  }

  return rv;
}

/**/

template <typename T>
inline std::string BytesToHex(const T& vch, bool fSpaces = false) {
  return BytesToHex(vch.begin(), vch.end(), fSpaces);
}

/********/

const int8_t p_util_hexdigit[256] = {
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  0,   1,   2,   3,
    4,  5,  6,  7,  8,  9,  -1, -1, -1, -1, -1, -1, -1, 0xa, 0xb, 0xc, 0xd, 0xe, 0xf, -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  0xa, 0xb, 0xc, 0xd, 0xe, 0xf, -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
    -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,  -1,
};

/**/

inline int8_t HexDigit(char c) {
  return p_util_hexdigit[static_cast<uint8_t>(c)];
}

/**/

inline std::vector<uint8_t> HexToBytes(const char* psz) {
  // convert hex dump to vector
  std::vector<uint8_t> vch;
  for (;;) {
    while (isspace(*psz) != 0) {
      psz++;
    }
    auto c = HexDigit(*psz++);
    if (c == static_cast<int8_t>(-1)) {
      break;
    }
    int8_t n = (c << 4);
    c = HexDigit(*psz++);
    if (c == static_cast<int8_t>(-1)) {
      break;
    }
    n |= c;
    vch.push_back(n);
  }
  return vch;
}

/**/

#endif


src/bip66.cpp
/********************************************************************************
*
*  The MIT License (MIT)
*
*  Copyright (c) Simon Downey <simon@ark.io> <https://github.com/sleepdefic1t>
*
*  Permission is hereby granted, free of charge, to any person obtaining a copy of
*  this software and associated documentation files (the "Software"), to deal in
*  the Software without restriction, including without limitation the rights to
*  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
*  the Software, and to permit persons to whom the Software is furnished to do so,
*  subject to the following conditions:
*
*  The above copyright notice and this permission notice shall be included in all
*  copies or substantial portions of the Software.
*
*  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
*  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
*  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
*  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
*  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
*  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
*
*  BIP66: https://github.com/bitcoin/bips/blob/master/bip-0066.mediawiki
*  adapted from: https://github.com/bitcoinjs/bip66/blob/master/index.js
*
*********************************************************************************/

#include "bip66.hpp"

#include <algorithm>
#include <array>
#include <cstdint>

////////////////////////////////////////////////////////////////////////////////
// Global Constants
constexpr uint8_t MIN_SIG_LENGTH        = 8U;
constexpr uint8_t MAX_SIG_LENGTH        = 72U;

constexpr uint8_t SIG_COMPOUND          = 0x30;

constexpr uint8_t ELEMENT_LENGTH        = 32U;

constexpr uint8_t ELEMENT_PADDING       = 0x00;
constexpr uint8_t ELEMENT_INTEGER       = 0x02;
constexpr uint8_t ELEMENT_NEGATIVE      = 0x80;

constexpr uint8_t SIG_SEQ_OFFSET        = 0U;
constexpr uint8_t SIG_LEN_OFFSET        = 1U;
constexpr uint8_t R_SEQ_OFFSET          = 2U;
constexpr uint8_t R_LEN_OFFSET          = 3U;
constexpr uint8_t R_BEGIN_OFFSET        = 4U;
constexpr uint8_t S_LEN_OFFSET          = 5U;  // + lenR
constexpr uint8_t S_BEGIN_OFFSET        = 6U;  // + lenR

////////////////////////////////////////////////////////////////////////////////
namespace bip66 {

////////////////////////////////////////////////////////////////////////////////
bool isValidElement(const uint8_t *element, const uint8_t &length) {
    // Check that the Element isn't too small or too large.
    if (length == 0U || length > ELEMENT_LENGTH + 1U) {
        return false;
    }

    // Check that the Element would not be interpreted as a negative value.
    if ((element[0] & ELEMENT_NEGATIVE) != 0U) {
        return false;
    }

    // Check that the Element is not excessively padded.
    if (length > 1U &&
        element[0] == ELEMENT_PADDING &&
        (element[1] & ELEMENT_NEGATIVE) == 0U) {
        return false;
    }

    return true;
}

////////////////////////////////////////////////////////////////////////////////
// Copy an Element adding or removing Padding if necessary.
//
// - Returns false if..
// .. the element is null.
// .. the destination is null.
// .. the element length is 0, or greater than 33.
//
// ---
// Copy Cases:
//
// Encoding and Element is Negative:
// - Prepend padding to the element and increment its length.
//
// Decoding and Element has Padding:
// - Skip the padding on copy.
//
// No Padding or Negative Elements detected:
// - We already checked the buffers and lengths, let's just go ahead and copy.
//
// ---
static bool copyElement(const uint8_t *element,
                        uint8_t *dst,
                        uint8_t &length,
                        bool encoding = true) {
    if (element == nullptr || dst == nullptr) {
        return false;
    }

    // Encoding and Element is Negative, add padding.
    if (encoding &&                                     // encoding
        length > 1U &&                                  // length > 1
        (element[0] & ELEMENT_NEGATIVE) != 0U) {        // 1st byte is negative
        dst[0] = ELEMENT_PADDING;
        std::copy_n(element, length, &dst[1]);
        ++length;
    }

    // Decoding and has padding, skip the padding.
    else if (!encoding &&                               // decoding
             element[0] == ELEMENT_PADDING &&           // has padding
             (element[1] & ELEMENT_NEGATIVE) != 0U) {   // 2nd byte is negative
        std::copy_n(&element[1], length - 1U, dst);

        // Padding is technically considered invalid,
        // let's check the original element instead.
        return isValidElement(element, length - 1U);
    }

    // No Padding or Negative Elements detected.
    else {
        std::copy_n(element, length, dst);
    }

    return isValidElement(dst, length);
}

////////////////////////////////////////////////////////////////////////////////
// bip66::encode
// 
//  Expects r and s to be positive DER integers.
//
//  The DER format uses the most significant bit as a sign bit (& 0x80).
//  If the significant bit is set AND the integer is positive, a 0x00 is prepended.
//
//  Examples:
//
//       0 =>     0x00
//       1 =>     0x01
//      -1 =>     0xff
//     127 =>     0x7f
//    -127 =>     0x81
//     128 =>   0x0080
//    -128 =>     0x80
//     255 =>   0x00ff
//    -255 =>   0xff01
//   16300 =>   0x3fac
//  -16300 =>   0xc054
//   62300 => 0x00f35c
//  -62300 => 0xff0ca4
//
// ---
//
// Format: 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S] [sighash]
//
// - total-length: 1-byte length descriptor of everything that follows,
//   excluding the sighash byte.
//
// - R-length: 1-byte length descriptor of the R value that follows.
//
// - R: arbitrary-length big-endian encoded R value. It must use the shortest
//   possible encoding for a positive integers (which means no null bytes at
//   the start, except a single one when the next byte has its highest bit set).
//
// - S-length: 1-byte length descriptor of the S value that follows.
//
// - S: arbitrary-length big-endian encoded S value. The same rules apply.
//
// --
bool encode(const uint8_t *rElement,
            const uint8_t &rElementLength,
            const uint8_t *sElement,
            const uint8_t &sElementLength,
            uint8_t *outSignature) {
    if (outSignature == nullptr) {
        return false;
    }

    std::array<std::array<uint8_t, ELEMENT_LENGTH + 2U> , 2U>  rs {};

    uint8_t lenR = rElementLength;
    uint8_t lenS = sElementLength;

    // Copy the Elements, checking for NULL
    if (!copyElement(rElement, rs.at(0).data(), lenR) ||
        !copyElement(sElement, rs.at(1).data(), lenS)) {
        return false;
    }

    const uint8_t outLen = S_BEGIN_OFFSET + lenR + lenS;

    if (outLen < MIN_SIG_LENGTH || outLen > MAX_SIG_LENGTH) {
        return false;
    }

    // 0x30 [total-length] 0x02 [R-length] [R] 0x02 [S-length] [S]
    outSignature[SIG_SEQ_OFFSET]    = SIG_COMPOUND;                 // 1 Byte
    outSignature[SIG_LEN_OFFSET]    = outLen - R_SEQ_OFFSET;        // 1 Byte

    // R Sequence
    outSignature[R_SEQ_OFFSET]      = ELEMENT_INTEGER;              // 1 Byte
    outSignature[R_LEN_OFFSET]      = lenR;                         // 1 Byte
    std::copy_n(rs.at(0).data(),                            // 32 <=> 33 Bytes
                lenR,
                &outSignature[R_BEGIN_OFFSET]);

    // S Sequence
    outSignature[R_BEGIN_OFFSET + lenR] = ELEMENT_INTEGER;          // 1 Byte
    outSignature[S_LEN_OFFSET + lenR]   = lenS;                     // 1 Byte
    std::copy_n(rs.at(1).data(),                            // 32 <=> 33 Bytes
                lenS,
                &outSignature[S_BEGIN_OFFSET + lenR]);

  return check(outSignature, outLen);                       // 70 <=> 72 Bytes
}
////////////////////////////////////////////////////////////////////////////////
// bip66::decode
//
// Decode the R and S elements from a Signature using Check validation.
bool decode(const uint8_t *signature,
            const uint8_t &signatureLength,
            uint8_t *outR,
            uint8_t *outS)  {
    if (!check(signature, signatureLength)) {
        return false;
    }

    uint8_t lenR = signature[R_LEN_OFFSET];
    uint8_t lenS = signature[S_LEN_OFFSET + lenR];

    return copyElement(&signature[R_BEGIN_OFFSET], outR, lenR, false) &&
           copyElement(&signature[S_BEGIN_OFFSET + lenR], outS, lenS, false);
}

////////////////////////////////////////////////////////////////////////////////
// bip66::check
//
// - Check that the Signature has valid DER encoding.
// - Ensure that each Element is..
// .. of the correct length corresponding to the Signature.
// .. a valid Element.
bool check(const uint8_t *signature, const uint8_t &length) {
    // Check the Signature Length
    if (length < MIN_SIG_LENGTH || length > MAX_SIG_LENGTH) {
        return false;
    }

    // Check for valid encoding.
    if (signature[SIG_SEQ_OFFSET] != SIG_COMPOUND ||
        signature[SIG_LEN_OFFSET] != length - R_SEQ_OFFSET) {
        return false;
    }

    uint8_t lenR = signature[R_LEN_OFFSET];

    // Check that both elements are integers.
    if (signature[R_SEQ_OFFSET] != ELEMENT_INTEGER ||
        signature[R_BEGIN_OFFSET + lenR] != ELEMENT_INTEGER) {
        return false;
    }

    uint8_t lenS = signature[S_LEN_OFFSET + lenR];

    // Check that the R & S lengths aren't too big.
    if (S_LEN_OFFSET + lenR >= length ||
        S_BEGIN_OFFSET + lenR + lenS != length) {
        return false;
    }

    if (signature[R_BEGIN_OFFSET + lenR] != ELEMENT_INTEGER) {
        return false;
    }

    return isValidElement(&signature[R_BEGIN_OFFSET], lenR) &&
           isValidElement(&signature[lenR + S_BEGIN_OFFSET], lenS);
}

}  // namespace bip66
